exercises/react/tests/react.rs

```diff
@@ -127,95 +156,96 @@ fn error_adding_callback_to_nonexistent_cell() {
 #[test]
 #[ignore]
 fn callbacks_only_fire_on_change() {
-    let mut values = Vec::new();
-    {
-        let mut reactor = Reactor::new();
-        let input = reactor.create_input(1);
-        let output = reactor.create_compute(&[input], |v| if v[0] < 3 { 111 } else { 222 }).unwrap();
-        assert!(reactor.add_callback(output, |v| values.push(v)).is_ok());
-        assert!(reactor.set_value(input, 2).is_ok());
-        assert!(reactor.set_value(input, 4).is_ok());
-    }
-    assert_eq!(values, vec![222]);
+    let cb = CallbackRecorder::new();
+    let mut reactor = Reactor::new();
+    let input = reactor.create_input(1);
+    let output = reactor.create_compute(&[input], |v| if v[0] < 3 { 111 } else { 222 }).unwrap();
+    assert!(reactor.add_callback(output, |v| cb.callback_called(v)).is_ok());
+
+    assert!(reactor.set_value(input, 2).is_ok());
+    cb.expect_not_to_have_been_called();
+    assert!(reactor.set_value(input, 4).is_ok());
+    cb.expect_to_have_been_called_with(222);
 }
 
 #[test]
 #[ignore]
 fn callbacks_can_be_added_and_removed() {
-    let mut values1 = Vec::new();
-    let mut values2 = Vec::new();
-    let mut values3 = Vec::new();
-    {
-        let mut reactor = Reactor::new();
-        let input = reactor.create_input(11);
-        let output = reactor.create_compute(&[input], |v| v[0] + 1).unwrap();
-        let callback = reactor.add_callback(output, |v| values1.push(v)).unwrap();
-        assert!(reactor.add_callback(output, |v| values2.push(v)).is_ok());
-        assert!(reactor.set_value(input, 31).is_ok());
-        assert!(reactor.remove_callback(output, callback).is_ok());
-        assert!(reactor.add_callback(output, |v| values3.push(v)).is_ok());
-        assert!(reactor.set_value(input, 41).is_ok());
-    }
-    assert_eq!(values1, vec![32]);
-    assert_eq!(values2, vec![32, 42]);
-    assert_eq!(values3, vec![42]);
+    let cb1 = CallbackRecorder::new();
+    let cb2 = CallbackRecorder::new();
+    let cb3 = CallbackRecorder::new();
+
+    let mut reactor = Reactor::new();
+    let input = reactor.create_input(11);
+    let output = reactor.create_compute(&[input], |v| v[0] + 1).unwrap();
+
+    let callback = reactor.add_callback(output, |v| cb1.callback_called(v)).unwrap();
+    assert!(reactor.add_callback(output, |v| cb2.callback_called(v)).is_ok());
+
+    assert!(reactor.set_value(input, 31).is_ok());
+    cb1.expect_to_have_been_called_with(32);
```

Comment:
Note as to how this applies to other languages:

Other languages would have been perfectly fine to use their language's equivalent of the old vector-based implementation, because they are free to check that the vector `== [32]` here. It was solely because I was unable to to do in Rust that I chose to use a Cell.

Other languages *need not* use Cell, but may still choose to if they feel it makes their tests more descriptive.

What all languages can indeed benefit from is some way of testing that the callbacks were called with value `32` (instead of waiting until the end to test this), no matter how that is implemented.